File: Extensions/Tessa.Extensions.Client/UI/PnrIncomingUIExtension.cs
using System;
using System.Threading.Tasks;
using System.Windows;
using Tessa.Cards;
using Tessa.Extensions.Shared.PnrCards;
using Tessa.Platform.Storage;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;
using Tessa.UI.Cards.Controls.AutoComplete;

namespace Tessa.Extensions.Client.UI
{
    public sealed class PnrIncomingUIExtension : CardUIExtension
    {
        private readonly ICardRepository cardRepository;
        private readonly Tessa.Platform.Runtime.ISession session;

        public PnrIncomingUIExtension(ICardRepository cardRepository, Tessa.Platform.Runtime.ISession session)
        {
            this.cardRepository = cardRepository;
            this.session = session;
        }
        
        private async Task<bool> GetIsUserInRole(Guid userID, Guid roleID)
        {
            CardRequest request = new CardRequest
            {
                RequestType = Shared.PnrRequestTypes.GetIsUserInRoleExtension,
                Info =
                    {
                        { "userID", userID },
                        { "roleID", roleID }
                    }
            };

            CardResponse response = await cardRepository.RequestAsync(request);

            return response.Info.Get<bool>("isUserInRole");
        }

        // установка видимости контролов, зависимых от Вида входящего документа
        private void SetDKDependenceVisibility(string documentKindIdx, ICardModel cardModel)
        {
            if (cardModel.Controls.TryGet("Correspondent", out IControlViewModel correspondent))
            {
                correspondent.ControlVisibility = documentKindIdx == "1-01" ? Visibility.Visible : Visibility.Collapsed;
                correspondent.Block.Rearrange();
            }

            IBlockViewModel block = cardModel.Blocks["ComplaintsBlock"];
            block.BlockVisibility = documentKindIdx == "1-02" ? Visibility.Visible : Visibility.Collapsed;
            block.Rearrange();

            IBlockViewModel IncomingLetterBlock = cardModel.Blocks["IncomingLetterBlock"];
            IncomingLetterBlock.BlockVisibility = documentKindIdx == PnrIncomingTypes.IncomingLetterIdx ? Visibility.Visible : Visibility.Collapsed;
            IncomingLetterBlock.Rearrange();
        }

        // установка значения скрытого поля Индекс ЮЛ
        async void SetLegalEntityIndex(AutoCompleteTableViewModel organizationControl, ICardModel cardModel)
        {
            if (organizationControl.Items.Count == 1)
            {
                RowAutoCompleteItem rowAutoCompleteItem = (RowAutoCompleteItem)organizationControl.Items[0];
                CardRow row = rowAutoCompleteItem.Row;
                Guid organizationID = (Guid)row["OrganizationID"];
                CardRequest request = new CardRequest
                {
                    RequestType = Shared.PnrRequestTypes.GetIndexLegalEntityRequestTypeID,
                    Info =
                    {
                        { "organizationID", organizationID }
                    }
                };

                CardResponse response = await cardRepository.RequestAsync(request);

                Tessa.Platform.Validation.ValidationResult result = response.ValidationResult.Build();
                TessaDialog.ShowNotEmpty(result);
                if (result.IsSuccessful)
                {
                    //cardModel.Card.Sections["PnrIncoming"].Fields["LegalEntityIndexID"] = response.Info.Get<Guid?>("LegalEntityID");
                    cardModel.Card.Sections["PnrIncoming"].Fields["LegalEntityIndexIdx"] = response.Info.Get<string>("LegalEntityIdx");
                }
            }
            else
            {
                //cardModel.Card.Sections["PnrIncoming"].Fields["LegalEntityIndexID"] = PnrLegalEntityIndex.WholeOrganizationID;
                cardModel.Card.Sections["PnrIncoming"].Fields["LegalEntityIndexIdx"] = PnrLegalEntityIndex.WholeOrganizationIdx;
            }
        }

        public override async Task Initialized(ICardUIExtensionContext context)
        {
            Card card = context.Card;
            ICardModel cardModel = context.Model;

            var currentUserID = session.User.ID;
            var isCurrentUserClerk = await GetIsUserInRole(currentUserID, PnrRoles.Clerk);
            var isCurrentUserOfficeManager = await GetIsUserInRole(currentUserID, PnrRoles.OfficeManager);
            var isCurrentUserAdmin = session.User.AccessLevel == Tessa.Platform.Runtime.UserAccessLevel.Administrator;

            if (cardModel.Controls.TryGet("DocumentKind", out IControlViewModel documentKind))
            {
                if (isCurrentUserAdmin || (isCurrentUserClerk && isCurrentUserOfficeManager))
                {
                    documentKind.IsReadOnly = false;
                }
                else
                {
                    documentKind.IsReadOnly = true;
                }
            }

            // Создание карточки
            if (card.StoreMode == CardStoreMode.Insert)
            {
                // Заполнить поле Дата регистрации текущей датой(можно просто продублировать значение из Дата создания)
                if (cardModel.Controls.TryGet("RegistrationDate", out IControlViewModel regDateControl))
                {
                    ((DateTimeViewModel)regDateControl).SelectedDate = DateTime.Now;
                }

                // поле Вид входящего документа автозополнено - устанавливаем видимость контролов в зависимости от Вида
                SetDKDependenceVisibility((string)card.Sections["PnrIncoming"].Fields["DocumentKindIdx"], cardModel);

            }
            else if (card.Sections["PnrIncoming"].Fields["DocumentKindID"] != null)
            // Открытие карточки. Есть Вид входящего документа - устанавливаем видимость контролов в зависимости от Вида
            {
                SetDKDependenceVisibility((string)card.Sections["PnrIncoming"].Fields["DocumentKindIdx"], cardModel);
            }

            // Организация ГК Пионер
            if (context.Model.Controls.TryGet("Organizations", out IControlViewModel organizations))
            {
                var autoComplete = (AutoCompleteTableViewModel)organizations;
                autoComplete.ValueSelected +=
                    async (sender, args) =>
                    {
                        SetLegalEntityIndex((AutoCompleteTableViewModel)sender, context.Model);
                    };
                autoComplete.ValueDeleted +=
                    async (sender, args) =>
                    {
                        SetLegalEntityIndex((AutoCompleteTableViewModel)sender, context.Model);
                    };
            }

            if ((Guid)card.Sections["PnrIncoming"].Fields["DocumentKindID"] == PnrIncomingTypes.IncomingComplaintsID &&
                card.Sections["PnrIncoming"].Fields["DepartmentID"] == null)
            {
                SetAutorDepartment(PnrIncomingTypes.IncomingComplaintsID, card);
            }

            context.Card.Sections["PnrIncoming"].FieldChanged += (s, e) =>
            {
                // Вид входящего документа: Visibility контролов
                if (e.FieldName == "DocumentKindIdx" && e.FieldValue != null)
                {
                    SetDKDependenceVisibility((string)e.FieldValue, cardModel);
                }

                if (e.FieldName == "DocumentKindID" && e.FieldValue != null)
                {
                    SetAutorDepartment((Guid)e.FieldValue, card);
                }
            };

            // return base.Initialized(context);
        }

        private async void SetAutorDepartment(Guid fieldValue, Card card)
        {
            if (fieldValue.Equals(PnrIncomingTypes.IncomingComplaintsID))
            {
                CardRequest request = new CardRequest
                {
                    RequestType = Shared.PnrRequestTypes.GetUserDepartmentInfoRequestTypeID,
                    Info =
                    {
                        { "authorID", card.CreatedByID }
                    }
                };

                CardResponse response = await cardRepository.RequestAsync(request);

                var result = response.ValidationResult.Build();
                if (result.IsSuccessful)
                {
                    card.Sections["PnrIncoming"].Fields["DepartmentID"] = response.Info.Get<Guid?>("DepartmentID");
                    card.Sections["PnrIncoming"].Fields["DepartmentName"] = response.Info.Get<string>("Name");
                    card.Sections["PnrIncoming"].Fields["DepartmentIdx"] = response.Info.Get<string>("Index");
                }
            }
            else
            {
                card.Sections["PnrIncoming"].Fields["DepartmentID"] = null;
                card.Sections["PnrIncoming"].Fields["DepartmentName"] = null;
                card.Sections["PnrIncoming"].Fields["DepartmentIdx"] = null;
            }
        }
    }
}

File: WebClient SDK/src/PnrScripts/UI/PnrIncomingUIExtension.ts
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { Guid, createTypedField, DotNetType, Visibility } from 'tessa/platform';
import { Card, CardStoreMode } from 'tessa/cards';
import { PnrCardTypes } from '../Shared/PnrCardTypes';
import { PnrIncomingTypes } from '../Shared/PnrIncomingTypes';
import { PnrRequestTypes } from '../Shared/PnrRequestTypes';
import { PnrLegalEntityIndex } from '../Shared/PnrLegalEntityIndex';
import { CardRequest, CardService } from 'tessa/cards/service';
import { AutoCompleteTableViewModel, RowAutoCompleteItem, DateTimeViewModel } from 'tessa/ui/cards/controls';
import Moment from 'moment';
import { userSession }  from 'common/utility';
import { PnrMetadataContainerClerk, PnrMetadataContainerOfficeManager } from 'src/PnrScripts/UI/PnrMetadataExtension';

export class PnrIncomingUIExtension extends CardUIExtension {
  public initialized(context: ICardUIExtensionContext) {
    if (!Guid.equals(context.card.typeId, PnrCardTypes.PnrIncomingTypeID)) {
      return;
    }
    let PnrIncoming = context.card.sections.tryGet("PnrIncoming");
    if (!PnrIncoming) return;

    const isCurrentUserAdmin = userSession.isAdmin;
    const isCurrentUserClerk = PnrMetadataContainerClerk.instance.isUserInRole;
    const isCurrentUserOfficeManager = PnrMetadataContainerOfficeManager.instance.isUserInRole;

    const documentKind = context.model.controls.get('DocumentKind');
    if (documentKind) {
      if (isCurrentUserAdmin || (isCurrentUserClerk && isCurrentUserOfficeManager)) {
        documentKind.isReadOnly = false;
      }
      else {
        documentKind.isReadOnly = true;
      }
    }

    // Создание карточки
    if (context.card.storeMode === CardStoreMode.Insert)
    {
      // Заполнить поле Дата регистрации текущей датой(можно просто продублировать значение из Дата создания)
      const regDateControl = context.model.controls.get('RegistrationDate') as DateTimeViewModel;
      if (regDateControl) {
        regDateControl.selectedDate = Moment(Date.now());
      }
      // поле Вид входящего документа автозополнено - устанавливаем видимость контролов в зависимости от Вида
      this.SetDKDependenceVisibility(PnrIncoming.fields.get('DocumentKindIdx'), context);
    }
    else if (PnrIncoming.fields.get('DocumentKindID') != null)
    {
      // Открытие карточки. Есть Вид входящего документа - устанавливаем видимость контролов в зависимости от Вида
      this.SetDKDependenceVisibility(PnrIncoming.fields.get('DocumentKindIdx'), context);
    }
    const organizations = context.model.controls.get('Organizations') as AutoCompleteTableViewModel;
    // Организация ГК Пионер
    if (organizations) {
      organizations.valueSet.add(async () => {
        await this.SetLegalEntityIndex(organizations, context);
      });
      organizations.valueDeleted.add(async () => {
        await this.SetLegalEntityIndex(organizations, context);
      });
    }

    if (PnrIncoming.fields.get('DocumentKindID') === PnrIncomingTypes.IncomingComplaintsID &&
     PnrIncoming.fields.get('DepartmentID') === null) {
      this.SetAutorDepartment(PnrIncomingTypes.IncomingComplaintsID, context.card);
    }

    PnrIncoming.fields.fieldChanged.add(async (e) => {
       // Вид входящего документа: Visibility контролов
       if (e.fieldName === "DocumentKindIdx" && e.fieldValue != null) {
          this.SetDKDependenceVisibility(e.fieldValue, context);
       }
       if (e.fieldName === "DocumentKindID" && e.fieldValue != null) {
          this.SetAutorDepartment(e.fieldValue, context.card);
       }
    });
  }

  private async SetAutorDepartment(fieldValue, card: Card) {
    let PnrIncoming = card.sections.tryGet("PnrIncoming");
        if (!PnrIncoming) return;
    if (fieldValue === PnrIncomingTypes.IncomingComplaintsID) {
        let request = new CardRequest();
        request.requestType = PnrRequestTypes.GetUserDepartmentInfoRequestTypeID,
        request.info["authorID"] = createTypedField(card.createdById, DotNetType.Guid);
        let response = await CardService.instance.request(request);
        if (response.validationResult.isSuccessful) {
          PnrIncoming.fields.set('DepartmentID', createTypedField(response.info.DepartmentID.$value, DotNetType.Guid));
          PnrIncoming.fields.set('DepartmentName', createTypedField(response.info.Name.$value, DotNetType.String));
          PnrIncoming.fields.set('DepartmentIdx', createTypedField(response.info.Index.$value, DotNetType.String));
        }
      }
    else {
        PnrIncoming.fields.set('DepartmentID', null);
        PnrIncoming.fields.set('DepartmentName', null);
        PnrIncoming.fields.set('DepartmentIdx', null);
    }
  }

  /** Установка видимости контролов, зависимых от Вида входящего документа */
  private SetDKDependenceVisibility(documentKindIdx, context: ICardUIExtensionContext)
  {
    const correspondent = context.model.controls.get('Correspondent');
    if (correspondent) {
        correspondent.controlVisibility = documentKindIdx === "1-01" ? Visibility.Visible : Visibility.Collapsed;
    }

    const block = context.model.blocks.get('ComplaintsBlock');
    if (block) {
      block.blockVisibility = documentKindIdx === "1-02" ? Visibility.Visible : Visibility.Collapsed;
    }

    const IncomingLetterBlock = context.model.blocks.get('IncomingLetterBlock');
    if (!IncomingLetterBlock) return;
    IncomingLetterBlock.blockVisibility = documentKindIdx === "1-01" ? Visibility.Visible : Visibility.Collapsed;
    }

  /** Установка значения скрытого поля Индекс ЮЛ */
  private async SetLegalEntityIndex(organizationControl: AutoCompleteTableViewModel, context: ICardUIExtensionContext)
  {
    if (organizationControl.items.length === 1) {
      let rowAutoCompleteItem = organizationControl.items[0] as RowAutoCompleteItem;
      let row = rowAutoCompleteItem.row;
      let organizationID = row.getField('OrganizationID');
      if (organizationID != null) {
        organizationID = organizationID.$value;
      }
      let LegalEntityIdx;
      let request = new CardRequest();
          request.requestType = PnrRequestTypes.GetIndexLegalEntityRequestTypeID;
          request.info["organizationID"] = createTypedField(organizationID, DotNetType.Guid);

      // проверка ответа
      let response = await CardService.instance.request(request);
      if (response.validationResult.isSuccessful) {
          LegalEntityIdx = response.info.LegalEntityIdx;
      }
      let PnrIncoming = context.card.sections.tryGet("PnrIncoming");
      if (LegalEntityIdx != null && PnrIncoming != null) {
        PnrIncoming.fields.set('LegalEntityIndexIdx', createTypedField(LegalEntityIdx.$value, DotNetType.String))
      }
    }
    else {
      let PnrIncoming = context.card.sections.tryGet("PnrIncoming");
      if (PnrIncoming != null) {
        PnrIncoming.fields.set('LegalEntityIndexIdx', createTypedField(PnrLegalEntityIndex.WholeOrganizationIdx, DotNetType.String))
      }
    }
  }
}